### PR TITLE
RPP -  Load payment methods through the request class

### LINKED ISSUE #6685: Payment Process: Load payment methods through the request class
### Description

The payment methods from https://github.com/Automattic/woocommerce-payments/issues/6682 and https://github.com/Automattic/woocommerce-payments/issues/6683 need to be instantiated before entering the payment process.

To address this issue, add a new `get_payment_method()` to the request class (issue https://github.com/Automattic/woocommerce-payments/issues/6684).

### Acceptance criteria

- The new method can provide either a saved, or a new payment method object.
- Use the legacy proxy from the DI set of issues to access `WC_Payment_Tokens`.

### Testing instructions

The new method should have 100% unit test coverage.

### Dev notes

Previously implemented as a payment method factory here: https://github.com/Automattic/woocommerce-payments/pull/6287/files#diff-5615624c5665b9bc368d784011b2e18ff0bab9cc3f00d6f414f51089119040ad. This could have changed within the gateway.

### Additional context

Epic: https://github.com/Automattic/woocommerce-payments/issues/6673
Project: paJDYF-9hL-p2

### PR DESCRIPTION
**Pull request by htdat**
Fixes #6685

#### Changes proposed in this Pull Request

- Add `get_payment_method` and its tests. 

#### Testing instructions

* Check tests and confirm they're all great.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `7100` or branch name `rpp/6685-load-payment-methods` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=rpp/6685-load-payment-methods) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 6aeaf6161907d1bde56b6a2604c83ab22ede4e9e
- Build time: 2023-08-31 12:35:13 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: src/Internal/Payment/PaymentRequest.php [hunk 1]
```diff
@@ -89,4 +106,42 @@ public function get_payment_method_id(): ?string {
...
+
+	/**
+	 * Gets payment method object from request.
+	 *
+	 * @throws PaymentRequestException
+	 */
+	public function get_payment_method(): PaymentMethodInterface {
+		$request = $this->request;
+
+		$is_woopayment_selected = isset( $request['payment_method'] ) && WC_Payment_Gateway_WCPay::GATEWAY_ID === $request['payment_method'];
```
Comment (RadoslavGeorgiev) [inline src/Internal/Payment/PaymentRequest.php:118]: Wrong operator here, should be `AND`.

File: src/Internal/Payment/PaymentRequest.php [hunk 2]
```diff
@@ -89,4 +106,42 @@ public function get_payment_method_id(): ?string {
...
+		if ( ! $is_woopayment_selected ) {
+			throw new PaymentRequestException( __( 'WooPayments is not used during checkout.', 'woocommerce-payments' ) );
+		}
+
+		if ( ! empty( $request['wcpay-payment-method'] ) ) {
```
Comment (RadoslavGeorgiev) [inline src/Internal/Payment/PaymentRequest.php]: Let's check whether the key exists first, to allow triggering notices.

More importantly, I think the logic here is a bit flawed: Even though the hidden field named `wcpay-payment-method` contains the tokenized payment method, and is only created/populated after the submission of the form, it is possible that the form could be re-submitted (after an error) with a new payment method entered, but a saved one selected. 

If you take a look at the existing `Payment_Information` class, generally both the new payment method is loaded (if available), and the selected token. However, the token takes precedence:

https://github.com/Automattic/woocommerce-payments/blob/1a9176f8104328aafc73c2f94ff06ca2d78d5155/includes/class-payment-information.php#L166-L172

File: src/Internal/Payment/PaymentRequest.php [hunk 3]
```diff
@@ -89,4 +106,42 @@ public function get_payment_method_id(): ?string {
...
+			$payment_method = sanitize_text_field( wp_unslash( $request['wcpay-payment-method'] ) );
+			return new NewPaymentMethod( $payment_method );
+		}
+
+		$token_param = 'wc-' . WC_Payment_Gateway_WCPay::GATEWAY_ID . '-payment-token';
+		if ( ! empty( $request[ $token_param ] ) ) {
+			$token_id = absint( wp_unslash( $request [ $token_param ] ) );
```
Comment (RadoslavGeorgiev) [inline src/Internal/Payment/PaymentRequest.php]: I have the feeling that you coded things this way intentionally, but the token ID could be `new` as well:

https://github.com/Automattic/woocommerce-payments/blob/1a9176f8104328aafc73c2f94ff06ca2d78d5155/includes/class-payment-information.php#L284-L295

If that happened here, `$token_id` would be `0`, and would bypass the `is_null( $token )` check.

File: src/Internal/Payment/PaymentRequest.php [hunk 4]
```diff
@@ -89,4 +106,42 @@ public function get_payment_method_id(): ?string {
...
+			 */
+			$token = $this->legacy_proxy->call_static( WC_Payment_Tokens::class, 'get', $token_id );
+
+			if ( is_null( $token ) ) {
+				throw new PaymentRequestException( __( 'Invalid saved payment method (token) ID', 'woocommerce-payments' ) );
+			}
+			return new SavedPaymentMethod( $token );
+		}
+
+		throw new PaymentRequestException( __( 'No valid payment method attached to the request.', 'woocommerce-payments' ) );
```
Comment (RadoslavGeorgiev) [inline src/Internal/Payment/PaymentRequest.php]: Let's make the message here a bit more user-friendly, avoiding `attached` or `request`. Something in the tone of `No payment method was selected` seems a bit more appropriate?

**Review by RadoslavGeorgiev — CHANGES_REQUESTED** [review]
Great start! I have a couple of early comments.

File: src/Internal/Payment/PaymentRequest.php (hunk 1 above)
Comment (htdat) [reply src/Internal/Payment/PaymentRequest.php:118]: I think I had fixed it in this commit 00eb0c0674837a5cea7a83370bab9a04ad2e3a47 before you commented here.

Currently, it's `&&` operator.

File: src/Internal/Payment/PaymentRequest.php (hunk 2 above)
Comment (htdat) [reply src/Internal/Payment/PaymentRequest.php]: > Let's check whether the key exists first, to allow triggering notices.

I think `empty` actually does both: check if the key exists, and ensure it's not an empty string `''`. https://3v4l.org/ugCT2

> More importantly, I think the logic here is a bit flawed:

This is very true! I fixed it here 74b0bc4b4a8892f0953eb0f6f3f1688f692d398a 
Actually, value of `wc-woocommerce_payments-payment-token` is `new`, and then `wcpay-payment-method` is attached when a new payment is added,

File: src/Internal/Payment/PaymentRequest.php (hunk 3 above)
Comment (htdat) [reply src/Internal/Payment/PaymentRequest.php]: Fixed here 74b0bc4b4a8892f0953eb0f6f3f1688f692d398a as mentioned above https://github.com/Automattic/woocommerce-payments/pull/7100#discussion_r1311036705

File: src/Internal/Payment/PaymentRequest.php (hunk 4 above)
Comment (htdat) [reply src/Internal/Payment/PaymentRequest.php]: Good point! Addressed d638f039122bd868581d082935237f94ea1a0e14

**Comment by htdat** [comment]
@RadoslavGeorgiev  - thanks for your detailed review. I've addressed all of them and ready to be review again.

File: src/Internal/Payment/PaymentRequest.php (hunk 1 above)
Comment (RadoslavGeorgiev) [reply src/Internal/Payment/PaymentRequest.php:118]: Oh, it was shortly after I `git pull`-ed last. Nice!

File: src/Internal/Payment/PaymentRequest.php (hunk 2 above)
Comment (RadoslavGeorgiev) [reply src/Internal/Payment/PaymentRequest.php]: > I think empty actually does both: check if the key exists, and ensure it's not an empty string ''. https://3v4l.org/ugCT2

Today I learned! https://www.php.net/manual/en/function.empty.php:

> empty() does not generate a warning if the variable does not exist.

File: tests/unit/src/Internal/Payment/PaymentRequestTest.php [hunk 5]
```diff
@@ -129,4 +140,87 @@ public function provider_test_order_id(): array {
...
+			$this->mock_legacy_proxy,
+			$request
+		);
+
+		$this->mock_legacy_proxy->expects( $this->once() )
+			->method( 'call_static' )
+			->with( WC_Payment_Tokens::class, 'get', 123456 )
+			->willReturn( $this->createMock( WC_Payment_Token::class ) );
+
+		$this->assertInstanceOf( SavedPaymentMethod::class, $this->sut->get_payment_method() );
```
Comment (RadoslavGeorgiev) [inline tests/unit/src/Internal/Payment/PaymentRequestTest.php]: In this method, and the next, one, let's verify that the returned object has the right ID/token contained within it. It's minor, but a piece of calm.

**Review by RadoslavGeorgiev — APPROVED** [review]
Changes look great, @dat! I have a small suggestion, feel free to address/ignore it, and ship the PR!

File: tests/unit/src/Internal/Payment/PaymentRequestTest.php (hunk 5 above)
Comment (htdat) [reply tests/unit/src/Internal/Payment/PaymentRequestTest.php]: Addressed 1dd1c4105433e1bdfebda498b1ab236db8261c2e

**Comment by htdat** [comment]
@RadoslavGeorgiev - thank you. I've addressed your last feedback. I am going to merge this PR tomorrow.

File: tests/unit/src/Internal/Payment/PaymentRequestTest.php [hunk 6]
```diff
@@ -21,12 +27,17 @@ class PaymentRequestTest extends WCPAY_UnitTestCase {
 	 */
 	private $sut;
 
+	public function setUp(): void {
+		parent::setUp();
+		$this->mock_legacy_proxy = $this->createMock( LegacyProxy::class );
```
Comment (RadoslavGeorgiev) [inline tests/unit/src/Internal/Payment/PaymentRequestTest.php:40]: Oh, I forgot about this, but had it as a local change, and remembered now: Let's define the `$mock_legacy_proxy` var with type `LegacyProxy|MockObject`. Not a problem functionally, but my editor raises too many alarms :D

Comment (htdat) [reply tests/unit/src/Internal/Payment/PaymentRequestTest.php:40]: Good point, I would also have declared property `$this->mock_legacy_proxy` too. Addressed  6aeaf6161907d1bde56b6a2604c83ab22ede4e9e

**Pull request merged**